SSA/WellTypedFramework.lean

```diff
@@ -18,118 +21,121 @@ inductive UserType (T : Type) : Type where
 
 namespace UserType
 
-instance: Inhabited (UserType BaseType) where default := UserType.unit
+instance: Inhabited (UserType β) where default := UserType.unit
 
-def toType [Godel BaseType] : UserType BaseType → Type
-  | .base t => Godel.toType t
+def toType [Godel β] : UserType β → Type
+  | .base t =>  ⟦t⟧
   | .pair k₁ k₂ => (toType k₁) × (toType k₂)
   | .triple k₁ k₂ k₃ => toType k₁ × toType k₂ × toType k₃
   | .unit => Unit
 
-def mkPair {BaseType : Type} [Godel BaseType] {k₁ k₂ : UserType BaseType}: toType k₁ → toType k₂ → toType (.pair k₁ k₂)
+instance [Godel β] : Godel (UserType β) where 
+  toType := toType 
+
+def mkPair [Godel β] {k₁ k₂ : UserType β}: ⟦k₁⟧ → ⟦k₂⟧ → ⟦k₁.pair k₂⟧
   | k₁, k₂ => (k₁, k₂)
 
-def mkTriple {BaseType : Type} [Godel BaseType] {k₁ k₂ k₃ : UserType BaseType}: toType k₁ → toType k₂ → toType k₃ → toType (.triple k₁ k₂ k₃)
+def mkTriple [Godel β] {k₁ k₂ k₃ : UserType β}: ⟦k₁⟧ → ⟦k₂⟧ → ⟦k₃⟧ → ⟦k₁.triple k₂ k₃⟧
   | k₁, k₂, k₃ => (k₁, k₂, k₃)
 
-def fstPair {BaseType : Type} [Godel BaseType] {k₁ k₂ : UserType BaseType} : toType (.pair k₁ k₂) → toType k₁
+def fstPair [Godel β] {k₁ k₂ : UserType β} : ⟦k₁.pair k₂⟧ → ⟦k₁⟧
   | (k₁, _) => k₁
 
-def sndPair {BaseType : Type} [Godel BaseType] {k₁ k₂ : UserType BaseType} : toType (.pair k₁ k₂) → toType k₂
+def sndPair [Godel β] {k₁ k₂ : UserType β} : ⟦k₁.pair k₂⟧ → ⟦k₂⟧
   | (_, k₂) => k₂
 
-def fstTriple {BaseType : Type} [Godel BaseType] {k₁ k₂ k₃ : UserType BaseType} : toType (.triple k₁ k₂ k₃) → toType k₁
+def fstTriple [Godel β] {k₁ k₂ k₃ : UserType β} : ⟦k₁.triple k₂ k₃⟧ → ⟦k₁⟧
   | (k₁, _, _) => k₁
 
-def sndTriple {BaseType : Type} [Godel BaseType] {k₁ k₂ k₃ : UserType BaseType} : toType (.triple k₁ k₂ k₃) → toType k₂
+def sndTriple [Godel β] {k₁ k₂ k₃ : UserType β} : ⟦k₁.triple k₂ k₃⟧ → ⟦k₂⟧
   | (_, k₂, _) => k₂
 
-def trdTriple {BaseType : Type} [Godel BaseType] {k₁ k₂ k₃ : UserType BaseType} : toType (.triple k₁ k₂ k₃) → toType k₃
+def trdTriple [Godel β] {k₁ k₂ k₃ : UserType β} : ⟦k₁.triple k₂ k₃⟧ → ⟦k₃⟧
   | (_, _, k₃) => k₃
 
 end UserType
 
 
-structure UserData {BaseType : Type} [Godel BaseType] where
-  type : UserType BaseType
-  value : type.toType
+structure UserData [Godel β] where
+  type : UserType β
+  value : toType type
 
 
--- @goens: I suggest removing [DecidableEq ...] [Godel ...]
```

Comment:
so @bollu style is removing comments :D? I was curious about the rationale (see comment on other commit)

Reply:
Heh, right, sorry for deleting the comment :) 

the rationale is from abstract algebra: you usually need no typeclass data to construct an object. It is thus evil to prevent users from constructing the object, since then your constructor is not at the maximum level of generality.

In terms of pure engineering, it reduces  the number of typeclasses to be deduced and makes error messages (and code) more concise :)